Register LLVM handlers for bad-alloc / OOM

cuviper opened the pull request:
LLVM's default bad-alloc handler may throw if exceptions are enabled,
and `operator new` isn't hooked at all by default. Now we register our
own handler that prints a message similar to fatal errors, then aborts.
We also call the function that registers the C++ `std::new_handler`.

Fixes #121305
Cc llvm/llvm-project#85281
r? @nikic

cuviper committed:
Register LLVM handlers for bad-alloc / OOM

LLVM's default bad-alloc handler may throw if exceptions are enabled,
and `operator new` isn't hooked at all by default. Now we register our
own handler that prints a message similar to fatal errors, then aborts.
We also call the function that registers the C++ `std::new_handler`.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8303258301/job/22727080498) [(plain)](https://github.com/rust-lang/rust/commit/0ade5a11f5705e33e031ae5b47c21fd0553d6675/checks/22727080498/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_ff37c15c-7fa8-4b5b-a58e-11e654f9defc
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=llvm-oom
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_ff37c15c-7fa8-4b5b-a58e-11e654f9defc
GITHUB_REF=refs/pull/122574/merge
GITHUB_REF_NAME=122574/merge
GITHUB_REF_PROTECTED=false
---
   Compiling fluent-bundle v0.15.2
   Compiling rustc_type_ir v0.0.0 (/checkout/compiler/rustc_type_ir)
   Compiling rustc_baked_icu_data v0.0.0 (/checkout/compiler/rustc_baked_icu_data)
warning: rustc_llvm@0.0.0: Compiler version doesn't include clang or GCC: "c++" "--version"
warning: rustc_llvm@0.0.0: llvm-wrapper/RustWrapper.cpp: In function ‘void BadAllocErrorHandler(void*, const char*, bool)’:
warning: rustc_llvm@0.0.0: llvm-wrapper/RustWrapper.cpp:106:8: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
warning: rustc_llvm@0.0.0:   106 |   write(2, OOM, strlen(OOM));
warning: rustc_llvm@0.0.0:       |   ~~~~~^~~~~~~~~~~~~~~~~~~~~
warning: rustc_llvm@0.0.0: llvm-wrapper/RustWrapper.cpp:107:8: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
warning: rustc_llvm@0.0.0:   107 |   write(2, Reason, strlen(Reason));
warning: rustc_llvm@0.0.0:       |   ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: rustc_llvm@0.0.0: llvm-wrapper/RustWrapper.cpp:108:8: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
warning: rustc_llvm@0.0.0:   108 |   write(2, "\n", 1);
   Compiling rustc_error_messages v0.0.0 (/checkout/compiler/rustc_error_messages)
   Compiling gsgdt v0.1.2
   Compiling ar_archive_writer v0.1.5
   Compiling thorin-dwp v0.7.0
---
##[group]Building stage1 library artifacts (x86_64-unknown-linux-gnu)
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names -Csymbol-mangling-version=v0 '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap)' '--check-cfg=cfg(stdarch_intel_sde)' '--check-cfg=cfg(no_fp_fmt_parse)' '--check-cfg=cfg(no_global_oom_handling)' '--check-cfg=cfg(no_rc)' '--check-cfg=cfg(no_sync)' '--check-cfg=cfg(netbsd10)' '--check-cfg=cfg(backtrace_in_libstd)' '--check-cfg=cfg(target_env,values("libnx","p2"))' '--check-cfg=cfg(target_arch,values("arm64ec","spirv","nvptx","xtensa"))' -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Cembed-bitcode=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' --target x86_64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 254)
  --- stderr
  rustc: /checkout/src/llvm-project/llvm/lib/Support/ErrorHandling.cpp:133: void llvm::install_bad_alloc_error_handler(fatal_error_handler_t, void *): Assertion `!ErrorHandler && "Bad alloc error handler already registered!\n"' failed.
  rustc exited with signal: 6 (SIGABRT) (core dumped)
{"llvm-bitcode-linker":"test-fail"}
##[group]Building bootstrap
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
##[endgroup]
ERROR: Tool `book` was not recorded in tool state.
ERROR: Tool `nomicon` was not recorded in tool state.
ERROR: Tool `reference` was not recorded in tool state.
ERROR: Tool `rust-by-example` was not recorded in tool state.
ERROR: Tool `edition-guide` was not recorded in tool state.
ERROR: Tool `embedded-book` was not recorded in tool state.
  local time: Fri Mar 15 23:15:52 UTC 2024
  network time: Fri, 15 Mar 2024 23:15:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

cuviper commented:
Oops, the bad-alloc installer had its assertion fixed in llvm/llvm-project#83160 -- but I think it will work if we just register that *before* the fatal error handler...

cuviper committed:
Aggressively ignore write errors during bad-alloc

cuviper committed:
Install the bad-alloc handler before fatal errors

The bad-alloc installer was incorrectly asserting that the other handler
isn't set yet, instead of checking its own, but we can avoid that by
changing the order we install them.

Ref: https://github.com/llvm/llvm-project/issues/83040

nikic commented:
@bors r+

bors commented:
:pushpin: Commit 8d374b1f2af876423435e47b66c01cd6fa38aaa1 has been approved by `nikic`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikic 8d374b1f2af876423435e47b66c01cd6fa38aaa1 -->
<!-- homu: {"type":"Approved","sha":"8d374b1f2af876423435e47b66c01cd6fa38aaa1","approver":"nikic","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 8d374b1f2af876423435e47b66c01cd6fa38aaa1 with merge 553506a5e2d01ad2183b5029543a2efe159e3563...
<!-- homu: {"type":"BuildStarted","head_sha":"8d374b1f2af876423435e47b66c01cd6fa38aaa1","merge_sha":"553506a5e2d01ad2183b5029543a2efe159e3563"} -->

rust-log-analyzer commented:
The job **`armhf-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8308093606/job/22737953793) [(plain)](https://github.com/rust-lang-ci/rust/commit/553506a5e2d01ad2183b5029543a2efe159e3563/checks/22737953793/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [ui] tests/ui/rfcs/rfc-2091-track-caller/std-panic-locations.rs#default stdout ----

error in revision `default`: test run failed!
command: RUST_TEST_THREADS="8" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/remote-test-client" "run" "0" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfcs/rfc-2091-track-caller/std-panic-locations.default/a"
--- stdout -------------------------------
uploaded "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfcs/rfc-2091-track-caller/std-panic-locations.default/a", waiting for result
------------------------------------------
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8308093606/job/22738253816)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8308093606/job/22738253816","builder_name":"checks-actions"} -->

cuviper commented:
"Connection reset" in remote-test-client

@bors retry
